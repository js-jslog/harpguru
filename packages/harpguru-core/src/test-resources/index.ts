export {
  inactiveCellsHarpStrata,
  activeCellsHarpStrata,
  buildMockUseGlobalImplementation,
} from './test-resources'
