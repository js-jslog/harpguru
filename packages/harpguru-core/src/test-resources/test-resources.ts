import { getHarpStrata } from 'harpstrata'
import type { HarpStrata, HarpStrataProps, ActiveIds } from 'harpstrata'
import {
  TuningIds,
  PitchIds,
  DegreeIds,
  PozitionIds,
  ValvingIds,
} from 'harpparts'

import {
  reduceFullMatrixToViewableMatrix,
  reduceViewableMatrixToLayoutFacts,
} from '../utils'
import { DisplayModes, ExperienceModes, FlushChannels } from '../types'
import type { ColumnBounds } from '../types'

const allActiveDegrees = [
  DegreeIds.Root,
  DegreeIds.Flat2,
  DegreeIds.Second,
  DegreeIds.Flat3,
  DegreeIds.Third,
  DegreeIds.Fourth,
  DegreeIds.Flat5,
  DegreeIds.Fifth,
  DegreeIds.Flat6,
  DegreeIds.Sixth,
  DegreeIds.Flat7,
  DegreeIds.Seventh,
]

const baseHarpStrataProps: HarpStrataProps = {
  tuningId: TuningIds.MajorDiatonic,
  valvingId: ValvingIds.NotValved,
  pozitionId: PozitionIds.First,
  harpKeyId: PitchIds.C,
  activeIds: [] as ActiveIds,
}
const inactiveCellsHarpStrataProps = baseHarpStrataProps
const activeCellsHarpStrataProps = {
  ...baseHarpStrataProps,
  activeIds: allActiveDegrees,
}

export const inactiveCellsHarpStrata = getHarpStrata(
  inactiveCellsHarpStrataProps
)
export const activeCellsHarpStrata = getHarpStrata(activeCellsHarpStrataProps)

type BuildMockUseGlobalImplementationProps = {
  readonly sourceHarpStrata?: HarpStrata
  readonly sourceColumnBounds?: ColumnBounds
  readonly experienceMode?: ExperienceModes
  readonly displayMode?: DisplayModes
  readonly bufferedActivityToggles?: ReadonlyArray<DegreeIds>
  readonly fragmentHarpFaceByOctaves?: boolean
  readonly flushChannel?: FlushChannels
  readonly activeQuizDegrees?: ReadonlyArray<DegreeIds>
}

type MockGlobalImplementation = (arg0: string) => unknown[] | undefined

export const buildMockUseGlobalImplementation = ({
  sourceHarpStrata = activeCellsHarpStrata,
  sourceColumnBounds = 'FIT',
  experienceMode = ExperienceModes.Explore,
  displayMode = DisplayModes.Degree,
  bufferedActivityToggles = [],
  fragmentHarpFaceByOctaves = true,
  flushChannel = FlushChannels.Regular,
  activeQuizDegrees = allActiveDegrees,
}: BuildMockUseGlobalImplementationProps): MockGlobalImplementation => {
  const mockGlobal = (stateItem: string) => {
    if (stateItem === 'activeDegreeMatrix')
      return [sourceHarpStrata.degreeMatrix]
    if (stateItem === 'activePitchMatrix') return [sourceHarpStrata.pitchMatrix]
    if (stateItem === 'activeInteractionMatrix')
      return [sourceHarpStrata.apparatus.interactionMatrix]
    if (stateItem === 'activeDegreeIds')
      return [sourceHarpStrata.activeDegreeIds]
    if (stateItem === 'activePitchIds') return [sourceHarpStrata.activePitchIds]
    if (stateItem === 'tuningId') return [sourceHarpStrata.apparatus.tuningId]
    if (stateItem === 'valvingId') return [sourceHarpStrata.apparatus.valvingId]
    if (stateItem === 'harpKeyId') return [sourceHarpStrata.harpKeyId]
    if (stateItem === 'pozitionId') return [sourceHarpStrata.pozitionId]
    if (stateItem === 'rootPitchId') return [sourceHarpStrata.rootPitchId]

    if (stateItem === 'columnBounds') return [sourceColumnBounds]
    const viewableDegreeMatrix = reduceFullMatrixToViewableMatrix(
      [[]] as const,
      sourceHarpStrata.degreeMatrix,
      sourceColumnBounds
    )
    const viewablePitchMatrix = reduceFullMatrixToViewableMatrix(
      [[]] as const,
      sourceHarpStrata.pitchMatrix,
      sourceColumnBounds
    )
    const viewableInteractionMatrix = reduceFullMatrixToViewableMatrix(
      [[]] as const,
      sourceHarpStrata.apparatus.interactionMatrix,
      sourceColumnBounds
    )
    if (stateItem === 'viewableDegreeMatrix') return [viewableDegreeMatrix]
    if (stateItem === 'viewablePitchMatrix') return [viewablePitchMatrix]
    if (stateItem === 'viewableInteractionMatrix')
      return [viewableInteractionMatrix]

    const layoutFacts = reduceViewableMatrixToLayoutFacts(
      viewableInteractionMatrix
    )
    if (stateItem === 'layoutFacts') return [layoutFacts]

    if (stateItem === 'activeExperienceMode') return [experienceMode]
    if (stateItem === 'activeDisplayMode') return [displayMode]
    if (stateItem === 'bufferedActivityToggles')
      return [bufferedActivityToggles]
    if (stateItem === 'fragmentHarpFaceByOctaves')
      return [fragmentHarpFaceByOctaves]
    if (stateItem === 'flushChannel') return [flushChannel]
    if (stateItem === 'activeQuizDegrees') return [activeQuizDegrees]
    return undefined
  }
  return mockGlobal
}
