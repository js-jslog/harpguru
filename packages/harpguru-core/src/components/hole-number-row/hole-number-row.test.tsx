import { useGlobal } from 'reactn'
import React from 'react'
import { render } from '@testing-library/react-native'

import { activeCellsHarpStrata } from '../../test-resources'

import { HoleNumberRow } from './hole-number-row'

jest.mock('reactn')
const mockUseGlobal = useGlobal as jest.Mock
mockUseGlobal.mockImplementation((stateItem: string) => {
  if (stateItem === 'activeHarpStrata') return [activeCellsHarpStrata]
  if (stateItem === 'fragmentHarpFaceByOctaves') return [true]
  if (stateItem === 'columnBounds') return ['FIT']
  return undefined
})

test('HoleNumberRow returns a 10 hole sequence', () => {
  const { queryByText, getByText } = render(
    <HoleNumberRow xRange={[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]} />
  )

  expect(queryByText('0')).toBeNull()
  expect(getByText('1')).toBeTruthy()
  expect(getByText('2')).toBeTruthy()
  expect(getByText('3')).toBeTruthy()
  expect(getByText('4')).toBeTruthy()
  expect(getByText('5')).toBeTruthy()
  expect(getByText('6')).toBeTruthy()
  expect(getByText('7')).toBeTruthy()
  expect(getByText('8')).toBeTruthy()
  expect(getByText('9')).toBeTruthy()
  expect(getByText('10')).toBeTruthy()
  expect(queryByText('11')).toBeNull()
})

test('A snapshot of HoleNumberRow', () => {
  const { toJSON } = render(
    <HoleNumberRow xRange={[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]} />
  )

  expect(toJSON()).toMatchSnapshot()
})
