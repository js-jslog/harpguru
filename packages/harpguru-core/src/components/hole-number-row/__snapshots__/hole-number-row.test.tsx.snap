// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A snapshot of HoleNumberRow 1`] = `
<View
  style={
    Object {
      "backgroundColor": "red",
      "flexDirection": "row",
      "height": 0,
      "justifyContent": "space-around",
      "zIndex": 3,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      1
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      2
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      3
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      4
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      5
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      6
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      7
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      8
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      9
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 21.120884235352566,
        "justifyContent": "center",
        "top": -10.560442117676283,
        "width": 55.295198637441516,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Object {
          "color": "#555",
          "fontSize": 21.120884235352566,
        }
      }
    >
      10
    </Text>
  </View>
</View>
`;
